File: 236. Lowest Common Ancestor of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool pathFinder(TreeNode* root, vector<TreeNode*>& n1, TreeNode* n){
        if(root == nullptr) return false;
        n1.push_back(root);
        if(root -> val == n -> val){
            return true;
        }

        if(pathFinder(root -> left, n1, n) || pathFinder(root -> right,n1,n)){
            return true;
        }

        n1.pop_back();
        return false;
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        vector<TreeNode*> p1, p2;
        pathFinder(root, p1, p);
        pathFinder(root, p2, q);

        int i = 0;
        while(i < p1.size() && i < p2.size()){
            if(p1[i] != p2[i]){
                return p1[i - 1];
            }
            i++;
        }
        return p1[i - 1];  
    }
};

File: 236. Lowest Common Ancestor of a Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    boolean flag = false;

    public boolean build(TreeNode root, ArrayList<TreeNode> path, TreeNode target) {
        if (root == null)
            return false;

        path.add(root);
        if (root == target)
            return true;

        if (build(root.left, path, target) || build(root.right, path, target)) {
            return true;
        }

        path.remove(path.size() - 1);
        return false;
    }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        ArrayList<TreeNode> path1 = new ArrayList<>();
        ArrayList<TreeNode> path2 = new ArrayList<>();
        build(root, path1, p);
        flag = false;
        build(root, path2, q);

        int n = path1.size();
        int m = path2.size();
        boolean check = true;
        int i = 0;
        while (i < path1.size() && i < path2.size() && path1.get(i) == path2.get(i)) {
            i++;
        }
        return path1.get(i - 1);
    }
}